documentation/specs/task-isolation-and-dependencies.md

```diff
@@ -0,0 +1,38 @@
+This problem is also described in microsoft/msbuild#1754.
+
+## Possible solution
+Use [`AssemblyLoadContext`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.2) (ALC) to provide binding isolation for tasks. Each task would be loaded into its own ALC instance.
+* The ALC would resolve all dependencies of the tasks (see dependency resolution below)
+* ALC would fallback to the Default for dependencies which the task doesn't carry with itself (Framework and so on)
+* ALC would probably have to forcefully fallback for MSBuild assemblies since by default lot of tasks will carry these, but the system requires for the MSBuild assemblies to be shared.
+We would probably also want to load groups of tasks which belong together into the same ALC (for example based on their location on disk) to improve performance. This will need some care as there's no guarantee that two random tasks have compatible dependency trees.
+
+## Potential risks
+* Has some small probability of causing breaks. Currently all assemblies from all tasks are loaded into the default context and thus are "visible" to everybody. Tasks with following properties might not work:
+  * Task has a dependency on an assembly, but it doesn't declare this dependency in its .deps.json and this dependency gets loaded through some other task. This is mostly fixable by implementing probing similar to today's behavior.
+  * Two tasks from different assemblies which somehow rely on sharing certain types. If the new system decides to load these in isolation they won't share types anymore and might not work.
+* Performance - task isolation inherently (and by design) leads to loading certain assemblies multiple times. This increases memory pressure and causes additional JITing and other related work.
+
+## Additional consideration
+* None of these changes would have any effect on MSBuild on .NET Framework
+* Task isolation alone could be achieved on existing MSBuild
+
+
+
+# Task dependency resolution
+## Problem definition
+Tasks with complex and specifically platform specific dependencies don't work out of the box. For example if a task uses [`LibGit2Sharp`](https://www.nuget.org/packages/LibGit2Sharp) package it will not work as is. `LibGit2Sharp` has native dependencies which are platform specific. While the package carries all of them, there's no built in support for the task to load the right ones. For example [source link](https://github.com/dotnet/sourcelink/blob/master/src/Microsoft.Build.Tasks.Git/GitLoaderContext.cs) runs into this problem.
+
+## Possible solution
+.NET Core uses `.deps.json` files to describe dependencies of components. It would be natural to treat task assemblies as components and use associated .deps.json file to determine their dependencies. This would make the system work nicely end to end with the .NET Core CLI/SDK and VS integration.
```

Comment:
What is the _authoring_ experience for `.deps.json` for "plugin" DLLs. What will a task author need to do on their end to make this work correctly?

Reply:
We generate one today for all .net core projects. But I filed https://github.com/dotnet/sdk/issues/2662

Reply:
As @nguerrera mentioned above, today it's enough to `dotnet new classlib` and `dotnet build` (or `dotnet publish`).